cohttp-curl-async/src/cohttp_curl_async.ml

```diff
@@ -0,0 +1,303 @@
+open Async_kernel
+module Time = Core_kernel.Time
+module Fd = Async_unix.Fd
+module Clock = Async_unix.Clock
+
+let ( let+ ) x f = Deferred.map x ~f
+
+module Sink = struct
+  type _ t = String : string t | Discard : unit t
+
+  let string = String
+  let discard = Discard
+end
+
+module Source = struct
+  type t = Empty | String of string
+
+  let empty = Empty
+  let string s = String s
+
+  let to_curl_callback t =
+    match t with
+    | Empty -> fun _ -> ""
+    | String s ->
+        let len = String.length s in
+        let pos = ref 0 in
+        fun max_asked ->
+          if !pos >= len then ""
+          else
+            let chunk_len = min (len - !pos) max_asked in
+            let res = String.sub s !pos chunk_len in
+            pos := !pos + chunk_len;
+            res
+end
+
+module Context = struct
+  type fd_events = {
+    fd : Fd.t;
+    mutable read : unit Ivar.t option;
+    mutable write : unit Ivar.t option;
+  }
+
+  type t = {
+    mt : Curl.Multi.mt;
+    wakeners : (Curl.t, Curl.curlCode Ivar.t) Hashtbl.t;
+    all_events : (Unix.file_descr, fd_events) Hashtbl.t;
+    mutable timeout : (unit, unit) Clock.Event.t option;
+  }
+
+  let create () =
+    let t =
+      {
+        mt = Curl.Multi.create ();
+        wakeners = Hashtbl.create 32;
+        all_events = Hashtbl.create 32;
+        timeout = None;
+      }
+    in
+    let rec finished s =
+      match Curl.Multi.remove_finished t.mt with
+      | None -> ()
+      | Some (h, code) ->
+          (match Hashtbl.find_opt t.wakeners h with
+          | None -> Printf.eprintf "curl_async(%s): orphan handle" s
```

Comment:
maybe use the async log module for this so it gets logged via the standard logging setup for async applications?

Reply:
I got rid of the logging.